edit-features-using-feature-forms/src/main/java/com/esri/arcgismaps/sample/editfeaturesusingfeatureforms/components/MapViewModel.kt

```diff
@@ -0,0 +1,230 @@
+import com.arcgismaps.data.ArcGISFeature
+import com.arcgismaps.data.ServiceFeatureTable
+import com.arcgismaps.exceptions.FeatureFormValidationException
+import com.arcgismaps.mapping.ArcGISMap
+import com.arcgismaps.mapping.PortalItem
+import com.arcgismaps.mapping.featureforms.FeatureForm
+import com.arcgismaps.mapping.featureforms.FeatureFormDefinition
+import com.arcgismaps.mapping.featureforms.FieldFormElement
+import com.arcgismaps.mapping.featureforms.FormElement
+import com.arcgismaps.mapping.featureforms.GroupFormElement
+import com.arcgismaps.mapping.layers.FeatureLayer
+import com.arcgismaps.mapping.view.SingleTapConfirmedEvent
+import com.arcgismaps.toolkit.geoviewcompose.MapViewProxy
+import com.esri.arcgismaps.sample.editfeaturesusingfeatureforms.R
+import com.esri.arcgismaps.sample.sampleslib.components.MessageDialogViewModel
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.launch
+
+/**
+ * A view model for the MainScreen UI
+ */
+class MapViewModel(application: Application) : AndroidViewModel(application) {
+
+    val mapViewProxy = MapViewProxy()
+
+    // feature forms enabled web-map showcasing places of interest with form fields
+    private var portalItem = PortalItem(application.getString(R.string.feature_form_web_map))
+
+    val map = ArcGISMap(portalItem)
+
+    // keep track of the selected feature form
+    private val _featureForm = MutableStateFlow<FeatureForm?>(null)
+    val featureForm: StateFlow<FeatureForm?> = _featureForm.asStateFlow()
+
+    // keep track of the list of validation errors
+    private val _errors = MutableStateFlow<List<ErrorInfo>>(listOf())
+    val errors: StateFlow<List<ErrorInfo>> = _errors.asStateFlow()
+
+    // create a ViewModel to handle dialog interactions
+    val messageDialogVM: MessageDialogViewModel = MessageDialogViewModel()
+
+    init {
+        viewModelScope.launch {
+            // load a map that has a FeatureFormDefinition on any of its layers
+            map.load()
+        }
+    }
+
+    /**
+     * Apply attribute edits to the Geodatabase backing the ServiceFeatureTable
+     * and refresh the local feature. Persisting changes to attributes is
+     * not part of the FeatureForm API.
+     *
+     * @param onEditsCompleted Invoked when edits are applied successfully
+     */
+    fun applyEdits(onEditsCompleted: () -> Unit) {
+        // build the list of errors
+        val errors = mutableListOf<ErrorInfo>()
+        val featureForm = _featureForm.value
+            ?: return messageDialogVM.showMessageDialog("Feature form state is not configured")
+
+        featureForm.validationErrors.value.forEach { entry ->
+            entry.value.forEach { error ->
+                featureForm.elements.getFormElement(entry.key)?.let { formElement ->
+                    if (formElement.isEditable.value || formElement.hasValueExpression) {
```

Comment:
This may require a comment, something like:
`If an element is editable or derives its value from an arcade expression, its errors must be corrected before submitting the form`.